include/SampleBuffer.h

```diff
@@ -25,333 +25,55 @@
-	{
-		m_frequency = freq;
-	}
-
-	inline void setSampleRate(sample_rate_t rate)
-	{
-		m_sampleRate = rate;
-	}
-
-	inline const sampleFrame * data() const
-	{
-		return m_data;
-	}
-
-	QString openAudioFile() const;
-	QString openAndSetAudioFile();
-	QString openAndSetWaveformFile();
-
-	QString & toBase64(QString & dst) const;
-
-
-	// protect calls from the GUI to this function with dataReadLock() and
-	// dataUnlock()
-	SampleBuffer * resample(const sample_rate_t srcSR, const sample_rate_t dstSR);
-
-	void normalizeSampleRate(const sample_rate_t srcSR, bool keepSettings = false);
-
-	// protect calls from the GUI to this function with dataReadLock() and
-	// dataUnlock(), out of loops for efficiency
-	inline sample_t userWaveSample(const float sample) const
-	{
-		f_cnt_t frames = m_frames;
-		sampleFrame * data = m_data;
-		const float frame = sample * frames;
-		f_cnt_t f1 = static_cast<f_cnt_t>(frame) % frames;
-		if (f1 < 0)
-		{
-			f1 += frames;
-		}
-		return linearInterpolate(data[f1][0], data[(f1 + 1) % frames][0], fraction(frame));
-	}
-
-	void dataReadLock()
-	{
-		m_varLock.lockForRead();
-	}
-
-	void dataUnlock()
-	{
-		m_varLock.unlock();
-	}
-
-
-	std::unique_ptr<OscillatorConstants::waveform_t> m_userAntiAliasWaveTable;
-
-
-public slots:
-	void setAudioFile(const QString & audioFile);
-	void loadFromBase64(const QString & data);
-	void setStartFrame(const lmms::f_cnt_t s);
-	void setEndFrame(const lmms::f_cnt_t e);
-	void setAmplification(float a);
-	void setReversed(bool on);
-	void sampleRateChanged();
+	using value_type = sampleFrame;
+	using reference = sampleFrame&;
+	using const_iterator = std::vector<sampleFrame>::const_iterator;
+	using const_reverse_iterator = std::vector<sampleFrame>::const_reverse_iterator;
+	using difference_type = std::vector<sampleFrame>::difference_type;
+	using size_type = std::vector<sampleFrame>::size_type;
+
+	SampleBuffer() = default;
+	explicit SampleBuffer(const QString& audioFile);
+	SampleBuffer(const QString& base64, int sampleRate);
+	SampleBuffer(std::vector<sampleFrame> data, int sampleRate);
+	SampleBuffer(
+		const sampleFrame* data, int numFrames, int sampleRate = Engine::audioEngine()->processingSampleRate());
+
+	friend void swap(SampleBuffer& first, SampleBuffer& second) noexcept;
+	auto toBase64() const -> QString;
+
+	inline auto audioFile() const -> const QString& { return m_audioFile; }
+	inline auto sampleRate() const -> sample_rate_t { return m_sampleRate; }
+
+	inline auto begin() const -> const_iterator { return m_data.begin(); }
+	inline auto end() const -> const_iterator { return m_data.end(); }
+	inline auto rbegin() const -> const_reverse_iterator { return m_data.rbegin(); }
+	inline auto rend() const -> const_reverse_iterator { return m_data.rend(); }
+
+	inline auto data() const -> const sampleFrame* { return m_data.data(); }
+	inline auto size() const -> size_type { return m_data.size(); }
+	inline auto empty() const -> bool { return m_data.empty(); }
```

Comment:
```suggestion
	auto audioFile() const -> const QString& { return m_audioFile; }
	auto sampleRate() const -> sample_rate_t { return m_sampleRate; }

	auto begin() const -> const_iterator { return m_data.begin(); }
	auto end() const -> const_iterator { return m_data.end(); }
	auto rbegin() const -> const_reverse_iterator { return m_data.rbegin(); }
	auto rend() const -> const_reverse_iterator { return m_data.rend(); }

	auto data() const -> const sampleFrame* { return m_data.data(); }
	auto size() const -> size_type { return m_data.size(); }
	auto empty() const -> bool { return m_data.empty(); }
```
According to [cppreference](https://en.cppreference.com/w/cpp/language/inline):
> A function defined entirely inside a [class/struct/union definition](https://en.cppreference.com/w/cpp/language/classes), whether it's a member function or a non-member friend function, is implicitly an inline function